include/hipSYCL/sycl/libkernel/group_functions.hpp

```diff
@@ -63,272 +62,377 @@
-  HIPSYCL_RETURN_DISPATCH_GROUP_ALGORITHM(__hipsycl_reduce_over_group, g, x,
-                                          binary_op);
+T reduce_over_group(Group g, T x, BinaryOperation binary_op) {
+  HIPSYCL_RETURN_DISPATCH_GROUP_ALGORITHM(__hipsycl_reduce_over_group, g, x, binary_op);
 }
 
-// exclusive_scan
-template <typename Group, typename InPtr, typename OutPtr, typename T,
-          typename BinaryOperation,
-          std::enable_if_t<is_group_v<std::decay_t<Group>>, bool> = true>
+template<class Group, typename T, typename BinaryOperation,
+    std::enable_if_t<detail::is_sp_group_v<std::decay_t<Group>>, bool> = true>
 HIPSYCL_BUILTIN
-OutPtr joint_exclusive_scan(Group g, InPtr first, InPtr last, OutPtr result, T init,
-                       BinaryOperation binary_op) {
-  HIPSYCL_RETURN_DISPATCH_GROUP_ALGORITHM(__hipsycl_joint_exclusive_scan, g,
-                                          first, last, result, init, binary_op);
+T reduce_over_group(
+    Group g, const detail::private_memory_access<T, Group> &x, BinaryOperation binary_op) {
+  HIPSYCL_RETURN_DISPATCH_GROUP_ALGORITHM(__hipsycl_reduce_over_group, g, x, binary_op);
 }
 
-template <typename Group, typename InPtr, typename OutPtr,
-          typename BinaryOperation,
-          std::enable_if_t<is_group_v<std::decay_t<Group>>, bool> = true>
+// exclusive_scan
+template<typename Group, typename InPtr, typename OutPtr, typename T, typename BinaryOperation,
+    std::enable_if_t<
+        (is_group_v<std::decay_t<Group>> || detail::is_sp_group_v<std::decay_t<Group>>), bool> =
+        true>
 HIPSYCL_BUILTIN
-OutPtr joint_exclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
-                            BinaryOperation binary_op) {
-  HIPSYCL_RETURN_DISPATCH_GROUP_ALGORITHM(__hipsycl_joint_exclusive_scan, g,
-                                          first, last, result, binary_op);
+OutPtr joint_exclusive_scan(
+    Group g, InPtr first, InPtr last, OutPtr result, T init, BinaryOperation binary_op) {
+  HIPSYCL_RETURN_DISPATCH_GROUP_ALGORITHM(
+      __hipsycl_joint_exclusive_scan, g, first, last, result, init, binary_op);
 }
 
-template<class Group, typename V, typename T, typename BinaryOperation>
+template<typename Group, typename V, typename T, typename BinaryOperation,
+    std::enable_if_t<is_group_v<std::decay_t<Group>>, bool> = true>
 HIPSYCL_BUILTIN
 T exclusive_scan_over_group(Group g, V x, T init, BinaryOperation binary_op) {
-  HIPSYCL_RETURN_DISPATCH_GROUP_ALGORITHM(__hipsycl_exclusive_scan_over_group,
-                                          g, x, init, binary_op);
-}
-
-template<typename Group, typename T, typename BinaryOperation,
-          std::enable_if_t<is_group_v<std::decay_t<Group>>, bool> = true>
-HIPSYCL_BUILTIN
-T exclusive_scan_over_group(Group g, T x, BinaryOperation binary_op) {
-  HIPSYCL_RETURN_DISPATCH_GROUP_ALGORITHM(__hipsycl_exclusive_scan_over_group,
-                                          g, x, binary_op);
```

Comment:
Where did `exclusive_scan_over_group` without `init` go? The spec says we need to have this overload. I don't see it anymore, am I missing something?

Reply:
I put all backend independent implementations in line 253+. Those are functions without init, which can be implemented using the identity and version with init. The exclusive scan without init for example is in line 378.